iml-request-retry/examples/demo-server-client.rs

```diff
@@ -0,0 +1,141 @@
+}
+
+fn main() {
+    let addr: SocketAddr = "127.0.0.1:8080".parse().unwrap();
+
+    let rt = Runtime::new().unwrap();
+    let executor = rt.executor();
+
+    // run server
+    executor.spawn(async move { server(&addr).await });
+
+    executor.spawn(async move {
+        let result = retry_client_obj(&addr, 1).await;
+        println!("retry_client_obj received {:?}", result);
+    });
+    executor.spawn(async move {
+        let result = retry_client_fun(&addr, 2).await;
+        println!("retry_client_fun received {:?}", result);
+    });
+
+    rt.block_on(async {
+        println!("Sleep for 30 seconds");
+        delay_for(Duration::from_secs(30)).await;
+        println!("End of sleep, shutting down");
+    });
+}
+
+async fn server(addr: &SocketAddr) {
+    let routes = warp::post().and(warp::body::json()).and_then(server_reply);
+    warp::serve(routes).run(*addr).await;
+}
+
+async fn retry_client_obj(addr: &SocketAddr, client_id: u32) -> Result<String, reqwest::Error> {
+    let mut policy_1 = |k: u32, e| match k {
+        0 => RetryAction::RetryNow,
+        k if k < 3 => RetryAction::WaitFor(Duration::from_secs((2 * k) as u64)),
+        _ => RetryAction::ReturnError(e),
+    };
+
+    retry_future(|| http_call(addr, client_id), &mut policy_1).await
+}
+
+async fn retry_client_fun(addr: &SocketAddr, client_id: u32) -> Result<String, reqwest::Error> {
+    fn policy_2<E: Debug>(k: u32, _e: E) -> RetryAction<E> {
+        if k == 0 {
+            RetryAction::RetryNow
+        } else {
+            RetryAction::WaitFor(Duration::from_secs(5))
+        }
+    }
+    retry_future_gen(|| http_call(addr, client_id), policy_2).await
+}
+
+async fn server_reply(mut hello: Hello) -> Result<Box<dyn warp::Reply>, warp::Rejection> {
+    GLOBAL_SERVER_COUNTER.fetch_add(1, Ordering::SeqCst);
+    let counter = GLOBAL_SERVER_COUNTER.load(Ordering::SeqCst);
+    if counter % 5 == 0 {
+        hello.counter = Some(counter);
+        println!(
+            "Incoming request from client={}, counter={}",
+            hello.client, counter
+        );
+        Ok(Box::new(warp::reply::json(&hello)) as Box<dyn warp::Reply>)
+    } else {
+        println!(
+            "Incoming request from client={}, counter={}, server is going to fail",
+            hello.client, counter
+        );
+        Ok(Box::new(StatusCode::INTERNAL_SERVER_ERROR) as Box<dyn warp::Reply>)
+    }
+}
+
+pub async fn http_call(addr: &SocketAddr, id: u32) -> reqwest::Result<String> {
+    let uri = Url::parse(&format!("http://{}", addr)).unwrap();
+
+    let mut map = HashMap::new();
+    map.insert("client", id);
+
+    // GET request with body
+    let client: Client = reqwest::Client::new();
+    let resp: Response = client.post(uri).json(&map).send().await?;
+
+    match resp.error_for_status() {
+        Ok(resp) => {
+            let resp_body = resp.text().await?.to_string();
+            Ok(resp_body)
+        }
+        Err(err) => Err(err),
+    }
```

Comment:
This can be simplified to:

`Ok(resp.error_for_status()?.text().await?.to_string())`